packages/components/src/date-time/date.js

```diff
@@ -2,29 +2,88 @@
  * External dependencies
  */
 import moment from 'moment';
+import classnames from 'classnames';
+
 // react-dates doesn't tree-shake correctly, so we import from the individual
 // component here, to avoid including too much of the library
 import DayPickerSingleDateController from 'react-dates/lib/components/DayPickerSingleDateController';
 
 /**
  * WordPress dependencies
  */
-import { Component, createRef } from '@wordpress/element';
-import { isRTL } from '@wordpress/i18n';
+import { Component, createRef, useEffect, useRef } from '@wordpress/element';
+import { isRTL, _n, sprintf } from '@wordpress/i18n';
 
 /**
  * Module Constants
  */
 const TIMEZONELESS_FORMAT = 'YYYY-MM-DDTHH:mm:ss';
+const ARIAL_LABEL_TIME_FORMAT = 'dddd, LL';
+
+function DatePickerDay( { day, events = [] } ) {
+	const ref = useRef();
+
+	/*
+	 * a11y hack to make the `There is/are n events` string
+	 * available speaking for readers,
+	 * re-defining the aria-label attribute.
+	 * This attribute is handled by the react-dates component.
+	 */
+	useEffect( () => {
+		// Bail when no parent node.
+		if ( ! ref?.current?.parentNode ) {
+			return;
+		}
+
+		const { parentNode } = ref.current;
+		const dayAriaLabel = moment( day ).format( ARIAL_LABEL_TIME_FORMAT );
+
+		if ( ! events.length ) {
+			// Set aria-label without event description.
+			parentNode.setAttribute( 'aria-label', dayAriaLabel );
+			return;
+		}
+
+		const dayWithEventsDescription = sprintf(
+			// translators: 1: Calendar day format, 2: Calendar event number.
+			_n(
+				'%1$s. There is %2$d event.',
+				'%1$s. There are %2$d events.',
+				events.length
+			),
+			dayAriaLabel,
+			events.length
+		);
+
+		parentNode.setAttribute( 'aria-label', dayWithEventsDescription );
+	}, [ ref, events.length ] );
```

Comment:
I don't really understand why we're accessing and modifying attributes using DOM APIs here. I'm pretty sure you can get React warnings when doing things like that to Dom elements controlled  by React.

Reply:
Also, using a "ref" as dependency is useless, a ref never changes its instance.

Reply:
> I don't really understand why we're accessing and modifying attributes using DOM APIs here. 

Because the component API doesn't support extending the aria-label attribute in a super flexible way. It's nice when setting the date format, which is great. We can use, and we do, the `dayAriaLabelFormat` prop:

```es6
<DayPickerSingleDateController
    dayAriaLabelFormat="dd"
/>
```

It generates the following markup (Saturday day):

```html
<td aria-label"Sa" ...>
```

Even is possible to set the property using arbitrary text, something like...

```es6
<DayPickerSingleDateController
    dayAriaLabelFormat="dd [🦊 retrofox]"
/>
```

<img width="337" alt="Screen Shot 2021-04-08 at 7 23 42 AM" src="https://user-images.githubusercontent.com/77539/114011276-6e8a9380-983b-11eb-8ce9-d10e46723142.png">

However, it isn't enough for our requirements, since what we'd like to achieve is setting a custom aria-label for each rendering day. Something like the following:

```es6
<DayPickerSingleDateController
	dayAriaLabelFormat={ ( day ) => {
		const dayAriaLabel = moment( day ).format( ARIAL_LABEL_TIME_FORMAT );
		const events = this.getEventsPerDay( day );

		return sprintf(
			// translators: 1: Calendar day format, 2: Calendar event number.
			_n(
				'%1$s. There is %2$d event.',
				'%1$s. There are %2$d events.',
				events.length
			),
			dayAriaLabel,
			events.length
		);
	} }

	// ....
/>
```

However, `dayAriaLabelFormat` only accepts a string. :'(
 
> I'm pretty sure you can get React warnings when doing things like that to Dom elements controlled by React.

I think everybody agrees with that. The approach is quite far away to be ideal, but it works and it shouldn't hurt in the sense the app shouldn't get broken. In the worse cases, it should perform any change in the markup.
Our issue is not being able to extend the component as we wish, and AFAIK, DOM  manipulation is the only solution without changing the react-dates component, something that [we researched](https://github.com/WordPress/gutenberg/pull/25428) a few months ago.

Reply:
I do expect this to break often though, React might restrict us from doing that. Do we have warnings now in the console because of that?

Reply:
All warnings that I see belong to the react-dates component. They already are in the trunk branch. I don't see any about the DOM manipulation process.

![image](https://user-images.githubusercontent.com/77539/114023864-c6c89200-9849-11eb-9b6f-3f9ece285c22.png)

![image](https://user-images.githubusercontent.com/77539/114024207-27f06580-984a-11eb-993f-3e6a8031c7c4.png)